containous/maesh
# ACL flag with reworked SMI/Service handling

### What does this PR do?

Closes #447 

This PR adds support for the ACL flag and rework the SMI/Service handling.

Currently, when the user installs Maesh he has to choose between 2 modes: Kubernetes service and SMI mode. Each of them provides different features:
- Kubernetes Service mode: Allow to use annotations on services (`circuit-breaker`, `rate-limiter` or `retry`).
- SMI mode: Allow to use SMI resources such as `TrafficSplit` to split the traffic coming to a service and `TrafficTarget` to authorize a set of pods to hit another set of pods.

This rewrite of the Provider will now allow us to use all these features at the same time, depending on if ACL mode is on or off:

 | Features              | ACL disabled | ACL enabled |
 |-----------------------|--------------|-------------|
 | Traffic-Type          | ✔            | ✔           |
 | Scheme                | ✔            | ✔           |
 | Retry                 | ✔            | ✔           |
 | Circuit-Breaker       | ✔            | ✔           |
 | Rate-Limit            | ✔            | ✔           |
 | Traffic-Split (SMI)   | ✔            | ✔           |
 | Traffic-Target (SMI)  | ✘            | ✔           |

As this is not intended to be a breaking change, the `smi` flag is still available. If the `acl` flag is not set but `smi` flag is set, it will behave like if the `acl` flag was set. 

### How to test it

- `make test`
- `make build`
- `docker tag containous/maesh:latest maesh-dev`

You can find some configuration files for different basic scenarios in the following archive:
[provider-test-assets.tar.gz](https://github.com/containous/maesh/files/4444460/provider-test-assets.tar.gz)

Test with ACL disabled:
- k3d create --server-arg --no-deploy --server-arg="traefik"`
- `k3d import-images maesh-dev`
- Install Maesh via Helm with the `acl=false` flag (or without it):
```
helm install maesh helm/chart/maesh \
  --namespace=maesh \
  --set controller.image.pullPolicy=IfNotPresent \
  --set controller.image.name=maesh-dev \
  --set controller.image.tag=latest \
  --set limits.tcp=20 \
  --set limits.http=20 \
  --set smi.deploy=true \
  --set acl=false
```
- `kubectl apply -f provider-test-assets/acl-disabled/http`

Test wit ACL enabled:
- Make sure you start from a clean state (issue #487).
- Install  Maesh via Helm with the `acl=true` flag:
```
helm install maesh helm/chart/maesh \
  --namespace=maesh \
  --set controller.image.pullPolicy=IfNotPresent \
  --set controller.image.name=maesh-dev \
  --set controller.image.tag=latest \
  --set limits.tcp=20 \
  --set limits.http=20 \
  --set smi.deploy=true \
  --set acl=true 
```
- `kubectl apply -f provider-test-assets/acl-enabled/http`

### Additional Notes

Co-authored-by: Landry Benguigui landry@containo.us
